src/compute/src/sink/persist_sink.rs

```diff
@@ -411,21 +416,16 @@ where
         // We go straight to the write handle to learn about the current upper
         // and upper advancements. The "persist_oks" stream is lying to us! For
         // example, when starting the persist_source with an as_of, the upper
-        // will jump to that as_of, and I believe REFRESH EVERY might also be
-        // playing tricks on us.
-        let gen_upper_future =
-            |current_upper: Antichain<Timestamp>,
-             mut handle: WriteHandle<SourceData, (), Timestamp, Diff>| {
-                let fut = async move {
-                    handle.wait_for_upper_past(&current_upper).await;
-                    let new_upper = handle.shared_upper();
-                    (handle, new_upper)
-                };
-
-                fut.boxed()
-            };
-
-        let mut upper_future = gen_upper_future(persist_frontier.clone(), write);
+        // will jump to that as_of.
+        let mut current_upper = persist_frontier.clone();
+        let upper_stream = async_stream::stream!({
+            loop {
+                write.wait_for_upper_past(&current_upper).await;
+                current_upper = write.shared_upper();
+                yield current_upper.clone()
+            }
+        });
```

Comment:
If the writer advances to the empty frontier, this loop will hang forever waiting for `wait_for_upper_past`, correct? And I think in this case we will never shut down the mint operator because it will never reach the `else` branch of the select.

I think this can be fixed by breaking out of the loop after yielding the empty frontier and then using a `Some(upper) = upper_stream.next() => { ... }` in the select below.